server/src/main/java/org/opensearch/wlm/QueryGroupService.java

```diff
@@ -8,32 +8,193 @@
+        Map<String, QueryGroupState> stateMap,
+        Set<QueryGroup> activeQueryGroups,
+        Set<QueryGroup> deletedQueryGroups
+    ) {
+        this.taskCancellationService = taskCancellationService;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.workloadManagementSettings = workloadManagementSettings;
+        this.nodeDuressTrackers = nodeDuressTrackers;
+        this.activeQueryGroups = activeQueryGroups;
+        this.deletedQueryGroups = deletedQueryGroups;
+        activeQueryGroups.forEach(queryGroup -> stateMap.putIfAbsent(queryGroup.get_id(), new QueryGroupState()));
+        this.queryGroupStateMap = stateMap;
+        this.queryGroupStateMap.put(QueryGroupTask.DEFAULT_QUERY_GROUP_ID_SUPPLIER.get(), new QueryGroupState());
+        taskCancellationService.setQueryGroupStateMapAccessor(this::getQueryGroupState);
+        clusterService.addStateApplier(this);
+    }
 
-    public QueryGroupService() {
-        this(new HashMap<>());
+    /**
+     * run at regular interval
+     */
+    protected void doRun() {
+        if (workloadManagementSettings.getWlmMode() == WlmMode.DISABLED) {
+            return;
+        }
+        taskCancellationService.refreshQueryGroups(activeQueryGroups, deletedQueryGroups);
+        taskCancellationService.cancelTasks(nodeDuressTrackers::isNodeInDuress);
+        taskCancellationService.pruneDeletedQueryGroups();
     }
 
-    public QueryGroupService(Map<String, QueryGroupState> queryGroupStateMap) {
-        this.queryGroupStateMap = queryGroupStateMap;
+    private QueryGroupState getQueryGroupState(final String queryGroupId) {
+        return queryGroupStateMap.get(queryGroupId);
+    }
+
+    /**
+     * {@link AbstractLifecycleComponent} lifecycle method
+     */
+    @Override
+    protected void doStart() {
+        scheduledFuture = threadPool.scheduleWithFixedDelay(() -> {
+            try {
+                doRun();
+            } catch (Exception e) {
+                logger.debug("Exception occurred in Query Sandbox service", e);
+            }
+        }, this.workloadManagementSettings.getQueryGroupServiceRunInterval(), ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel();
+        }
+    }
+
+    @Override
+    protected void doClose() throws IOException {}
+
+    @Override
+    public void applyClusterState(ClusterChangedEvent event) {
+        // Retrieve the current and previous cluster states
+        Metadata previousMetadata = event.previousState().metadata();
+        Metadata currentMetadata = event.state().metadata();
+
+        // Extract the query groups from both the current and previous cluster states
+        Map<String, QueryGroup> previousQueryGroups = previousMetadata.queryGroups();
+        Map<String, QueryGroup> currentQueryGroups = currentMetadata.queryGroups();
```

Comment:
In general we are looking to avoid taking dependencies on cluster state since updates are single threaded and doesn't allow the cluster to scale beyond a certain point. For my understanding is this something that can be maintained elsewhere

Reply:
Do you mean the query groups ?

Reply:
@Bukhtawar - Thanks for bringing this up. While we can think about other ways of solving this problem, more generally, I am wondering why the `ClusterStateAppliers` and `ClusterStateListeners` need to be single threaded. While consuming single `ClusterStateEvent`, we can execute all the appliers and listeners in parallel, given there is no implicit guarantee on the order of these appliers and listeners anyway. For example:

```
for (ClusterStateApplier applier : clusterStateAppliers) {
            logger.trace("calling [{}] with change to version [{}]", applier, clusterChangedEvent.state().version());
            try (TimingHandle ignored = stopWatch.timing("running applier [" + applier + "]")) {
                long applierStartTimeNS = System.nanoTime();
                // Execute all the appliers in parallel using separate threadpool
                applier.applyClusterState(clusterChangedEvent);
            }
// wait for all appliers to finish executing to prevent cluster state update thread to consume next event
// record latency for the cluster state applier operation
}
```

We can do same thing for listeners as well